File: design_patterns/C++/Structural/decorator.cxx
#include <iostream>
#include <string>

class Coffee {
  public:
	virtual int getCost() = 0;
	virtual std::string getDescription() = 0;
};

class SimpleCoffee: public Coffee {
  public:
	int getCost() {
		return 10;
	}

	std::string getDescription() {
		return "Simple coffee";
	}
};

class MilkCoffee: public Coffee {
  protected:
	Coffee* coffee = nullptr;

  public:
	MilkCoffee(Coffee* coffee) {
		this->coffee = coffee;
	}

	int getCost() {
		return this->coffee->getCost() + 2;
	}

	std::string getDescription() {
		return this->coffee->getDescription() + ", milk";
	}
};

class WhipCoffee: public Coffee {
  protected:
	Coffee* coffee = nullptr;

  public:
	WhipCoffee(Coffee* coffee) {
		this->coffee = coffee;
	}

	int getCost() {
		return this->coffee->getCost() + 2;
	}

	std::string getDescription() {
		return this->coffee->getDescription() + ", wilp";
	}
};

class VanillaCoffee: public Coffee {
  protected:
	Coffee* coffee = nullptr;

  public:
	VanillaCoffee(Coffee* coffee) {
		this->coffee = coffee;
	}

	int getCost() {
		return this->coffee->getCost() + 2;
	}

	std::string getDescription() {
		return this->coffee->getDescription() + ", milk";
	}
};

int main() {
	auto someCoffee1 = new SimpleCoffee();
	std::cout << someCoffee1->getCost() << std::endl;
	std::cout << someCoffee1->getDescription() << std::endl;
	// =============
	// 10
	// Simple coffee
	// =============
	auto someCoffee2 = new MilkCoffee(someCoffee1);
	std::cout << someCoffee2->getCost() << std::endl;
	std::cout << someCoffee2->getDescription() << std::endl;
	// ===================
	// 12
	// Simple coffee, milk
	// ===================
	auto someCoffee3 = new WhipCoffee(someCoffee2);
	std::cout << someCoffee3->getCost() << std::endl;
	std::cout << someCoffee3->getDescription() << std::endl;
	// =========================
	// 14
	// Simple coffee, milk, wilp
	// =========================
	auto someCoffee4 = new VanillaCoffee(someCoffee3);
	std::cout << someCoffee4->getCost() << std::endl;
	std::cout << someCoffee4->getDescription() << std::endl;
	// ===============================
	// 16
	// Simple coffee, milk, wilp, milk
	// ===============================
	return 0;
}

File: design_patterns/Python/Structural/decorator.py
class Coffee:

    def get_cost(self):
        raise NotImplementedError

    def get_description(self):
        raise NotImplementedError


class SimpleCoffee(Coffee):

    def get_cost(self):
        return 10

    def get_description(self):
        return 'Simple coffee'


class MilkCoffee(Coffee):

    def __init__(self, coffee):
        self.coffee = coffee

    def get_cost(self):
        return self.coffee.get_cost() + 2

    def get_description(self):
        return self.coffee.get_description() + ', milk'


class WhipCoffee(Coffee):

    def __init__(self, coffee):
        self.coffee = coffee

    def get_cost(self):
        return self.coffee.get_cost() + 5

    def get_description(self):
        return self.coffee.get_description() + ', whip'


class VanillaCoffee(Coffee):

    def __init__(self, coffee):
        self.coffee = coffee

    def get_cost(self):
        return self.coffee.get_cost() + 3

    def get_description(self):
        return self.coffee.get_description() + ', vanilla'


if __name__ == '__main__':

    coffee = SimpleCoffee()
    assert coffee.get_cost() == 10
    assert coffee.get_description() == 'Simple coffee'

    coffee = MilkCoffee(coffee)
    assert coffee.get_cost() == 12
    assert coffee.get_description() == 'Simple coffee, milk'

    coffee = WhipCoffee(coffee)
    assert coffee.get_cost() == 17
    assert coffee.get_description() == 'Simple coffee, milk, whip'

    coffee = VanillaCoffee(coffee)
    assert coffee.get_cost() == 20
    assert coffee.get_description() == 'Simple coffee, milk, whip, vanilla'
